### PR TITLE
fix issue #3704 regarding integer attribute values being cast to decimal

### PR DESCRIPTION
**Pull request by digitalpianism**

### CODE REVIEW

File: app/code/Magento/Eav/Model/Entity/Collection/AbstractCollection.php [hunk 1]
```diff
@@ -1175,7 +1175,11 @@ public function _loadAttributes($printQuery = false, $logQuery = false)
         foreach ($selectGroups as $selects) {
             if (!empty($selects)) {
                 try {
-                    $select = implode(' UNION ALL ', $selects);
+                    if (is_array($selects))
+                    {
+                        $select = implode(' UNION ALL ', $selects);
+                    }
+                    else $select = $selects;
```
Comment (antonkril) [inline app/code/Magento/Eav/Model/Entity/Collection/AbstractCollection.php]: please fix code style:

```
if (..) {
} else {
}
```

**Comment by davidalger** [comment]
@digitalpianism Thank you for your contribution. Please accept the contributors license agreement so this PR can be further processed. This can be done by clicking the "Details" link next to the "license/cla" check below.

**Comment by digitalpianism** [comment]
@antonkril I've provided a quick test to show how to replicate the bug in the issue: https://github.com/magento/magento2/issues/3704

**Comment by digitalpianism** [comment]
@davidalger I've done that about 3 different times on 2 different browsers but it doesn't seem to work even if it tells me "thanks for signing CLA, we're redirecting you to github"

I've tried revoking the OAuth access from my account but still doesn't work

**Comment by digitalpianism** [comment]
@davidalger ok done, the email address I'm using to commit code had not been verified my bad ;)

**Comment by vkublytskyi** [comment]
@digitalpianism Thank you for so deep investigation of the issue and provided fix. But please pay attention that Magento\Eav\Model\ResourceModel\Helper::getLoadAttributesSelectGroups() is also used in  Magento\Eav\Model\Entity\AbstractEntity so it should be also modified as Magento\Eav\Model\Entity\Collection\AbstractCollection.

However this fix may decrease performance as count of requests to DB will be increased. To improve performance Magento\Eav\Model\ResourceModel\Helper::getLoadAttributesSelectGroups() may combine selects for varchar, text and datetime in one group.

Also as issue appears only in case when only decimal and integers attributes are present due to MySQL casting logic for union statement fix without performance degradation may be provided. For this we need to use "CAST(value AS char)" expression.

**Comment by digitalpianism** [comment]
@vkublytskyi thanks for your comment, I'm about to make another commit regarding the `Magento\Eav\Model\Entity\AbstractEntity`. 

I agree with you regarding the performance issue, my main goal was to fix the bug, I did not focus on how to refactor the framework to get that bug fix optimized, I guess there's people more experienced than me to do so.

**Comment by vkublytskyi** [comment]
@digitalpianism Thanks for your contribution and for catching this bug.
We will investigate performance degradation and apply improvements if needed. Internal ticket to merge pull request and fix possible performance issues is MAGETWO-50319

**Comment by piotrekkaminski** [comment]
The issue is being processed now.

**Pull request merged**